src/transaction/transaction.rs

```diff
@@ -499,16 +503,28 @@ impl Transaction {
     /// # });
     /// ```
     pub async fn commit(&mut self) -> Result<Option<Timestamp>> {
-        if !matches!(
-            self.status,
-            TransactionStatus::StartedCommit | TransactionStatus::Active
-        ) {
-            return Err(Error::OperationAfterCommitError);
+        {
+            let mut status = self.status.write().await;
+            if !matches!(
+                *status,
+                TransactionStatus::StartedCommit | TransactionStatus::Active
+            ) {
+                return Err(Error::OperationAfterCommitError);
+            }
+            *status = TransactionStatus::StartedCommit;
         }
-        self.status = TransactionStatus::StartedCommit;
 
         let primary_key = self.buffer.get_primary_key().await;
         let mutations = self.buffer.to_proto_mutations().await;
+        if self.options.kind == TransactionKind::Optimistic
+            && self.options.heartbeat
+            && self.spawn_heartbeat
+        {
+            self.spawn_heartbeat = false;
+            if !mutations.is_empty() {
+                self.spawn_heartbeat().await;
+            }
```

Comment:
There is a same check in `commit()`. Maybe we can move this check out, and early returns if there is no need to commit/send heartbeat.